presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java

```diff
@@ -0,0 +1,195 @@
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+
+            OperatorContext operatorContext = driverContext.addOperatorContext(
+                    operatorId,
+                    planNodeId,
+                    AssignUniqueIdOperator.class.getSimpleName());
+            return new AssignUniqueIdOperator(operatorContext, types, valuePool);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new AssignUniqueIdOperatorFactory(operatorId, planNodeId, types);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private boolean finishing;
+    private final AtomicLong valuePool;
+    private final List<Type> types;
+    private final LongValueCompressor valueCompressor = new LongValueCompressor();
+
+    private Page inputPage;
+    private long counter;
+    private long lastCounterValue;
+
+    public AssignUniqueIdOperator(
+            OperatorContext operatorContext,
+            List<Type> types,
+            AtomicLong valuePool)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = ImmutableList.copyOf(types);
+        this.valuePool = requireNonNull(valuePool, "valuePool is null");
+
+        TaskId fullTaskId = operatorContext.getDriverContext().getTaskId();
+
+        valueCompressor.store(fullTaskId.getStageId().getId(), 10)
+                .store(fullTaskId.getId(), 14);
+
+        requestValues();
+    }
+
+    private void requestValues()
+    {
+        lastCounterValue = valuePool.addAndGet(VALUES_PER_REQUEST);
+        counter = lastCounterValue - VALUES_PER_REQUEST;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && inputPage == null;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && inputPage == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        requireNonNull(page, "page is null");
+        checkState(inputPage == null);
+        inputPage = page;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (inputPage == null) {
+            return null;
+        }
+
+        Page outputPage = getRowsWithUniqueColumns();
+        inputPage = null;
+        return outputPage;
+    }
+
+    private Page getRowsWithUniqueColumns()
+    {
+        int inputPageChannelCount = inputPage.getChannelCount();
```

Comment:
It should be possible to determine this value "statically" for the operator based on how it's configured at creation time.